File: solutions/c/bob/1/bob.c
#include "bob.h"

/*
    Question?           -> Sure.
    YELLING             -> Whoa, chill out!
    YELL A QUESTION?    -> Calm down, I know what I'm doing!
    silence             -> Fine. Be that way!
    anything else       -> Whatever.
*/

char *hey_bob(char *greeting){
    int len = strlen(greeting);
    // Are you yelling
    bool isYelling = false;
    for(int i = 0; i < len; i++) if(isupper(greeting[i])) isYelling = true; // Find at least 1 uppercase
    bool isSilence = true;
    // Are you asking a question?
    bool isaQuestion = false;
    char *questionMark = strchr(greeting, '?');
    if(questionMark) {
        isaQuestion = true;
        while(*questionMark != '\0'){
            // If it finds a character after the ? it is not a question
            if(isalnum((unsigned char)*questionMark)) {
                isaQuestion = false;
                break;
            }
            questionMark++;
        }
    }
    for(int i = 0; i < len-1; i++){
        if(greeting[i] != ' ' && greeting[i] != '\t' && greeting[i] != '\r' && greeting[i] != '\n'){
            isSilence = false; // Not silence if it is any character other than ' ', '\t', '\r', '\n'
            if(islower(greeting[i])){
                isYelling = false; // Not yelling if there is a lowercase
            }
        }
    }
    if(isSilence)
        return "Fine. Be that way!";
    if(isYelling && isaQuestion)
        return "Calm down, I know what I'm doing!";
    else if(isYelling)
        return "Whoa, chill out!";
    else if(isaQuestion)
        return "Sure.";
    else return "Whatever.";
    return NULL;
}

File: solutions/zig/bob/1/bob.zig
const std = @import("std");
const ascii = std.ascii;

pub fn response(s: []const u8) []const u8 {
    var hasLetter = false;
    var hasLower = false;
    var lastNonSpace:u8 = 0;

    for(s) |c|{
        if(!ascii.isWhitespace(c)) lastNonSpace = c;
        if(ascii.isAlphabetic(c)){
            hasLetter = true;
            if(ascii.isLower(c)){
                hasLower = true;
            }
        }
    }

    const silence = (lastNonSpace == 0);
    const question = (lastNonSpace == '?');
    const yelling = (hasLetter and !hasLower);

    if (silence) {
        return "Fine. Be that way!";
    } else if (question and yelling) {
        return "Calm down, I know what I'm doing!";
    } else if (question) {
        return "Sure.";
    } else if (yelling) {
        return "Whoa, chill out!";
    } else {
        return "Whatever.";
    }
}
